# Trabalho T1 - Erros em Aproximações numéricas

### MTM 224 - Métodos Numéricos Computacionais
### Eng. Química(305) - Turma 14

*Nomes*: David Disconzi, 201620715; Gabriel Righi, 201612819; Júlia Pesarico, 20182064.



---
## Questão 1
Determinar o valor de x que satisfaz a equação.

In [11]:
import math

def frac_molar():
    err = 10.0
    x0 = 0
    x1 = 0.05
    count = 0

    f = lambda x: (x/(x-1)) * (math.sqrt(6)/(2+x)) - 0.05
    
    while err> 0.0001:
        x = abs((x0*f(x1) - x1*f(x0))/(f(x1)-f(x0)))   
        err = abs(x-x1)/abs(x)      
        x0 = x1
        x1 = x
    
        print (count, "ªIteração -> Valor de x = %.9f"%x)
        print ("\t\tErro: ", err)
        count = count + 1
        
frac_molar()

0 ªIteração -> Valor de x = 0.039753177
		Erro:  0.2577611002737755
1 ªIteração -> Valor de x = 0.037811002
		Erro:  0.05136534847070694
2 ªIteração -> Valor de x = 0.038357994
		Erro:  0.014260190325714546
3 ªIteração -> Valor de x = 0.038418498
		Erro:  0.0015748486855598318
4 ªIteração -> Valor de x = 0.038392003
		Erro:  0.0006900991460630069
5 ªIteração -> Valor de x = 0.038390518
		Erro:  3.8683729629490414e-05


###### Comentario
Como vimos pelo código, utilizando o método das Secantes e com uma precisão de ε = ${10}^{-4}$ com poucas iterações foi possível encontrar o valor desejado.

---
## Questão 2
Determinar o valor de h utilizando o método do ponto-fixo com precisão ε = ${10}^{-5}$

In [7]:
import math

def volume():
    r = 1.0
    V = 0.5
    #valor inicial de h definido como 1
    h = 1.0
    h_ant = h
    err = 10
    #contador de iterações
    count = 0;
    f = lambda h: ((math.pi*math.pow(h,2) * (3*r - h))/3)*V
    while(err > math.pow(10, -5)):
        h_ant = h
        h = f(h)
        err = abs(h - h_ant)/abs(h)
        count = count + 1
        
        print("----------------------------- \nNúmero da iteração: ", count)
        print("Valor de h: ", h)
        print("Erro: ", err)

    print("--------------------------")        
    print("\n Final do laço. Valor de h:", h)
    
    
volume()

----------------------------- 
Número da iteração:  1
Valor de h:  1.0471975511965976
Erro:  0.04507034144862788
----------------------------- 
Número da iteração:  2
Valor de h:  1.1212802412882463
Erro:  0.06606973650631229
----------------------------- 
Número da iteração:  3
Valor de h:  1.2367700616187771
Erro:  0.09338018756644956
----------------------------- 
Número da iteração:  4
Valor de h:  1.41216518743116
Erro:  0.1242029809072411
----------------------------- 
Número da iteração:  5
Valor de h:  1.6579634182898768
Erro:  0.14825310869177552
----------------------------- 
Número da iteração:  6
Valor de h:  1.931580731043027
Erro:  0.14165460876460528
----------------------------- 
Número da iteração:  7
Valor de h:  2.087209596666246
Erro:  0.07456312287553386
----------------------------- 
Número da iteração:  8
Valor de h:  2.082101099987655
Erro:  0.002453529599797687
----------------------------- 
Número da iteração:  9
Valor de h:  2.083517242659806
Erro:  0.0006796

###### Comentario:
Após 14 iterações utilizando o método do ponto-fixo, atingimos o erro solicitado, e o valor final de h encontrado foi: 2.0832174714375227

---
## Questão 3
Determinar o valor de h1 utilizando o método da bisseção

In [2]:
import math


def Volume(R,r,h):
    return(math.pi*h/3)*(math.pow(R,2) + math.pow(r,2) + R*r)
    
def Emp(y):
    return 200*Volume(1,0.5,1)-1000*Volume(1,(0.5+y-0.5*y),(1-y))

def bissecao():
    
    a_bissecao = 0.0
    b_bissecao = 1.0
    err_bissecao = 10.0
    iterator = 0
    
    while err_bissecao > 0.000001:
        x_medio = (a_bissecao + b_bissecao)/2
    
        if Emp(a_bissecao)*Emp(x_medio) < 0:
            b_bissecao = x_medio
        else:
            a_bissecao = x_medio
        
        err_bissecao = abs(Emp(x_medio))
        iterator = iterator + 1
    
    print("Numero de Iterações necessárias: ", iterator, " -> h1 = ", x_medio)
    
def secante():
    
    limite = 0.0001  
    a_secante =0
    b_secante =1
    err_secante = 10.0
    count = 0
    
    while(err_secante> limite):
        aux = b_secante - ((b_secante - a_secante)/(Emp(b_secante) - Emp(a_secante)))*Emp(b_secante)
        a_secante = b_secante
        b_secante =  aux
        err_secante = abs(aux - a_secante)/abs(aux)
        print ("Iteração: ", count, " x = ", aux, " erro = ", err_secante)
        count = count + 1


print("Metodo da Bissecao:")     
bissecao()
print("\nMetodo da Secante:")
secante()
    

Metodo da Bissecao:
Numero de Iterações necessárias:  29  -> h1 =  0.875777455046773

Metodo da Secante:
Iteração:  0  x =  0.8  erro =  0.24999999999999994
Iteração:  1  x =  0.8708487084870848  erro =  0.08135593220338976
Iteração:  2  x =  0.8759824097788258  erro =  0.00586050728237471
Iteração:  3  x =  0.8757769159284954  erro =  0.00023464177531164364
Iteração:  4  x =  0.8757774553079667  erro =  6.158864538476374e-07


###### Comentario
Como é possível perceber, a diferença do número de iterações necessárias para atingir o valor esperado é impactante se comparado o Método da Bissecao com o Método das Secantes. Assim, enquanto que o método da Bissecao levou 30 iterações dentro do laço While, o Método das Secantes com 5 iterações conseguiu o mesmo resultado com a precisão necessária. Logo, é válido afirmar que o Método das Secantes para este caso, se torna mais viável, pois não consome muito tempo de processamento da máquina.

---
## Questão 4
Determinar o ângulo $\theta$ de forma que a área S da seção seja um terço da área do círculo utilizando os métodos da bisseção e do ponto-fixo

In [4]:
import math

def sec_circular():
    
    #definição de variáveis
    a=90    
    b=180
    err = 10
    x_ant = 0
    count = 0
    
    while err > 0.0001:
        x = (a+b)/2
        result = (math.sin(x) + ((2*math.pi)/3)) * (math.sin(a) + ((2*math.pi)/3))
        if result < 0:
            b = x
        else:
            a = x
            err = abs(x - x_ant)/abs(x)
            x_ant=x
        
        print(count, "ªIteração -> Valor de x = ", x)
        count = count + 1
            
    print("\nValor do angulo theta: ", x)
    
sec_circular()


0 ºIteração -> Valor de x =  135.0
1 ºIteração -> Valor de x =  157.5
2 ºIteração -> Valor de x =  168.75
3 ºIteração -> Valor de x =  174.375
4 ºIteração -> Valor de x =  177.1875
5 ºIteração -> Valor de x =  178.59375
6 ºIteração -> Valor de x =  179.296875
7 ºIteração -> Valor de x =  179.6484375
8 ºIteração -> Valor de x =  179.82421875
9 ºIteração -> Valor de x =  179.912109375
10 ºIteração -> Valor de x =  179.9560546875
11 ºIteração -> Valor de x =  179.97802734375
12 ºIteração -> Valor de x =  179.989013671875

Valor do angulo theta:  179.989013671875


###### Comentario
Como é possível perceber, após 13 iteações finalmente o programa encontrou o valor do ângulo theta com a precisão desejada, que no caso foi setada para ε = ${10}^{-4}$

---
## Questão 5
Determinar o valor da temperatura final da equação

In [1]:
import math

def termodinamica():
    
    #Definição de valores para as variáveis:
    R = 8.314
    A = 1.702
    B = 9.081 * (math.pow(10,-3))
    C = -2.164 * (math.pow(10,-6))
    n = 2.0
    Ti = 300.0
    Q = 20000.0
    
    f = lambda x: n*R*((A*(x - Ti)) + ((B/2)*(math.pow(x,2) - math.pow(Ti,2))+((C/3)*(math.pow(x,3) - math.pow(Ti,3)))))-Q
    
    err = 10.0
    a = 500
    b = 600
    x_medio = 1.0
    x_ant = 1.0
    count = 0
    
    while err > 0.001:
        x_medio = ((a+b)/2)
        if((f(a) * f(x_medio)) < 0):
            b = x_medio
        else:
            a = x_medio
        print("Iteração ", count, ": x = ", x_medio)
        err = abs(x_medio - x_ant)/abs(x_medio)
        x_ant = x_medio
        count = count+1
        
    print("\nResultado final:\nNumero de iterações:", count-1, "\nValor de Tf:", x_medio)
    
termodinamica()

Iteração  0 : x =  550.0
Iteração  1 : x =  525.0
Iteração  2 : x =  537.5
Iteração  3 : x =  531.25
Iteração  4 : x =  534.375
Iteração  5 : x =  535.9375
Iteração  6 : x =  535.15625
Iteração  7 : x =  535.546875

Resultado final:
Numero de iterações: 7 
Valor de Tf: 535.546875


###### Comentário
Usando o método da biseção, com a=500 e b=600 precisamos de 7 iterações para chegar na solução com erro menor que 0,001. Como é possível perceber, o valor da temperatura final Tf é 535.546875.